### PR TITLE
Error Handling overhaul.

### PR DESCRIPTION
**Pull request by daschl**
error handling! @simonbasle @ingenthr

### CODE REVIEW

**Comment by daschl** [comment]
No idea why this conflicts its rebased on doc-reorg master at time of PR :/

File: content/sdk/java/handling-error-conditions.dita [hunk 1]
```diff
@@ -3,6 +3,908 @@
 <topic id="topic_bpn_4bv_xv">
   <title>Handling Errors</title>
   <body>
-    <p>TBD</p>
+    <p>All Couchbase specific exceptions are derived from <apiname>CouchbaseException</apiname>. All the other
+  Exception types used are part of the JDK, including <apiname>IllegalArgumentException</apiname>, <apiname>TimeoutException</apiname> and others.</p>
+  <p>The following section explains of the exception types.</p>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: dangling "of". Maybe rephrase to  "The following section explains the exception types used by the SDK."?

File: content/sdk/java/handling-error-conditions.dita [hunk 2]
```diff
@@ -3,6 +3,908 @@
...
+                                <codeph>Observable</codeph> emits an error or the max count is reached. If the
+                              count is reached, the Observable will not be resubscribed, but the error is
+                              propagated down the stream. This operator is recommended for use.</li>
+                            <li><apiname>retry(Func2&lt;Integer, Throwable, Boolean&gt; predicate)</apiname>:
+                              Instantly retry as long as the predicate returns true. Arguments to the predicate
+                              are the number of tries, as well as the exception type. </li>
+                          </ul><p>Since the <codeph>predicate</codeph> method provides the most flexibility, it
+                            is recommended to use it. If you only want to handle a specific exception and retry
+                            a maximum of <codeph>MAX_TRIES</codeph> times, you can do it like this:
+                            </p><codeblock outputclass="language-java">   Observable .error(new CASMismatchException()) .retry(new \
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: formatting problem here, better format and align the codeblock and remove the \ characters (not sure if it needs a CDATA but it might)

File: content/sdk/java/handling-error-conditions.dita [hunk 3]
```diff
@@ -3,6 +3,908 @@
...
+                      instanceof CASMismatchException) &amp;&amp; tries &lt;MAX_TRIES; } }) .subscribe();         </codeblock><p>Try
+                            replacing <codeph>CASMismatchException</codeph> with something else and you will see
+                            that it does not try to retry, but rather propagates the error downstream. You can
+                            use this technique to handle specific errors differently by adding more retry
+                            operators in the pipeline. </p><p>Using the <codeph>retry</codeph> with
+                              <codeph>predicate</codeph> also allows you to log the number of retries for a
+                            specific error. If you use the <apiname>doOnError</apiname> for logging, it's harder
+                            to log the number of retries.</p><p>The synchronous equivalent to the latest code
+                            looks like
+                          this:</p><codeblock outputclass="language-java">   int tries = 0; while(true) { tries++; try { pretendWorkThatMaybeThrows(); // \
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: same formatting problem as above, with \ at end of lines

Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: adding a general comment

**Comment by simonbasle** [comment]
general comment: all codeblock should be revisited and properly realigned and reformatted. In particular, the lines following the opening <codeblock> must be aligned and indented relative to the start of the line, not the <codeblock>. There is a lot of codeblocks with dangling / and badly split lines, weirdly (didn't check if it was in the original source).

File: content/sdk/java/handling-error-conditions.dita [hunk 4]
```diff
@@ -3,6 +3,908 @@
...
+            </tgroup>
+          </table>
+
+    </section>
+
+    <section>
+      <title>Connection Management Errors</title>
+
+
+      <p>If you are trying to open a <apiname>Bucket</apiname> and it does not exist you get a <apiname>BucketDoesNotExistException</apiname>. If it has a different password compared to the supplied one you get a <apiname>InvalidPasswordException</apiname> instead.</p>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: should we say that starting with Watson, `InvalidPasswordException` is not thrown (because the server "hides" this information)

File: content/sdk/java/handling-error-conditions.dita [hunk 5]
```diff
@@ -3,6 +3,908 @@
...
+                              <entry>yes</entry>
+                              <entry>The server reports a temporary failure and it is very likely to be lock-related (like an already locked key or a bad cas used for unlock).</entry>
+                          </row>
+                        </tbody>
+                      </tgroup>
+                    </table>
+
+
+
+        <p>During a View request the following exception can happen:</p>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: I would mention that for Views, you not only want to watch out for exceptions but also for the `error()` method (and then link to the view doc if this is where we explain `isSuccess()` and `error()` in more details?)

File: content/sdk/java/handling-error-conditions.dita [hunk 6]
```diff
@@ -3,6 +3,908 @@
...
+                          <row>
+                              <entry><codeph>ViewDoesNotExistException</codeph></entry>
+                              <entry>no</entry>
+                              <entry>The view does not exist on the server (or the design document where it should belong) or is not published.</entry>
+                          </row>
+                        </tbody>
+                      </tgroup>
+                    </table>
+
+        <p>Similarly, when a N1QL request is issued watch out for the following:</p>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: I would mention that for N1QL, you not only want to watch out for exceptions but also for the `finalSuccess()` and `errors()` methods (and then link to the n1ql doc if this is where we want to explain these in more details?)

File: content/sdk/java/handling-error-conditions.dita [hunk 7]
```diff
@@ -3,6 +3,908 @@
...
+
+          <p>Here is the gotcha: this synchronous example only works great if the fallback is static. If you
+              need to fallback into another database call, for example, you quickly get into nested
+              error handling and timeouts are a pain to handle (since they start to accumulate for every
+              synchronous call). It is recommended to use asynchronous fallbacks and then block at the
+              very end through <apiname>toBlocking().single()</apiname> or equivalents.</p>
+
+    </section>
+
+    <section>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: maybe put all these reference sections before the "how to deal with (async) errors" sections?

File: content/sdk/java/handling-error-conditions.dita [hunk 8]
```diff
@@ -3,6 +3,908 @@
...
+              </tbody>
+            </tgroup>
+          </table>
+
+    </section>
+
+    <section>
+        <title>Conceptual Error Types</title>
+    </section>
+    <section conref="../shared/couchbase-errors.dita#toplevel/begin" conrefend="../shared/couchbase-errors.dita#toplevel/end"/>
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: I would also pull that up and above the "reference" sections

**Comment by ingenthr** [comment]
I'll hold off on merging this until @daschl has responded to the reviews.  I can't assign it to him though.  Please assign it to me when done.

File: content/sdk/java/handling-error-conditions.dita (hunk 3 above)
Comment (daschl) [inline content/sdk/java/handling-error-conditions.dita]: yeah the badly formatted stuff somehow carried over, I'll revisit all of them

File: content/sdk/java/handling-error-conditions.dita (hunk 4 above)
Comment (daschl) [inline content/sdk/java/handling-error-conditions.dita]: I think we still throw, just checked :)

Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: my bad, meant `BucketDoesntExistException`

**Comment by daschl** [comment]
Alright, huge update @ingenthr @simonbasle

File: content/sdk/java/handling-error-conditions.dita [hunk 9]
```diff
@@ -3,6 +3,975 @@
...
+
+     <p>In the async case, errors are events like every other for your subscribers. Once an error
+               happens, your <codeph>Subscriber</codeph> is notified in the method
+                 <apiname>onError(Throwable)</apiname>, and you can handle it the way you want to. Note
+               that by <codeph>Observable</codeph> contract, after the <apiname>onError</apiname> event,
+               no more <apiname>onNext</apiname> events will happen.</p>
+
+<codeblock outputclass="language-java"><![CDATA[Observable
+   .error(new Exception("I'm failing"))
+   .subscribe(new Subscriber&lt;Object&gt;() {
```
Comment (simonbasle) [inline content/sdk/java/handling-error-conditions.dita]: typo here, `&lt;` and `&gt;` in a CDATA gets rendered as is instead of `<` and `>`

Comment (ingenthr) [inline content/sdk/java/handling-error-conditions.dita]: Fixed those and others.

**Comment by ingenthr** [comment]
Rebased, reviewed, fixed a few things and merged.

@daschl: please add a shortdesc for all topics and if you can, please spell check.  I found a number of clearly misspelled words or missing parts of words, even in code.
